src/handlers/bulk/processing/handler.js

```diff
@@ -0,0 +1,337 @@
+ optionally within square brackets <email>.
+
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ * Georgi Georgiev <georgi.georgiev@modusbox.com>
+ --------------
+ ******/
+'use strict'
+
+const Logger = require('@mojaloop/central-services-shared').Logger
+const BulkTransferService = require('../../../domain/bulkTransfer')
+const Util = require('../../lib/utility')
+const Kafka = require('../../lib/kafka')
+// const Validator = require('../shared/validator')
+const Enum = require('../../../lib/enum')
+const TransferEventType = Enum.transferEventType
+const TransferEventAction = Enum.transferEventAction
+const Errors = require('../../../lib/errors')
+const errorType = Errors.errorType
+const Metrics = require('@mojaloop/central-services-metrics')
+const Config = require('../../../lib/config')
+const Mongoose = require('../../../lib/mongodb').Mongoose
+const decodePayload = require('@mojaloop/central-services-stream').Kafka.Protocol.decodePayload
+const { BulkTransferResponseModel } = require('../../../models/mongo/bulkTransfer')
+
+const location = { module: 'BulkProcessingHandler', method: '', path: '' } // var object used as pointer
+
+const consumerCommit = true
+const fromSwitch = true
+
+// const processingHandlerMessageProtocol = {
+//   value: {
+//     id: null,
+//     from: null,
+//     to: null,
+//     type: 'application/json',
+//     content: {
+//       headers: null,
+//       payload: null
+//     },
+//     metadata: {
+//       event: {
+//         id: Uuid(),
+//         responseTo: 'dfa',
+//         type: 'transfer',
+//         action: 'bulk-prepare',
+//         createdAt: null,
+//         state: {
+//           status: 'success',
+//           code: 0
+//         }
+//       }
+//     }
+//   }
+// }
+
+const connectMongoose = async () => {
+  let db = await Mongoose.connect(Config.MONGODB_URI, {
+    promiseLibrary: global.Promise
+  })
+  return db
+}
+
+/**
+ * @function BulkProcessingHandler
+ *
+ * @async
+ * @description This is the consumer callback function that gets registered to a topic. This then gets a list of messages,
+ * we will only ever use the first message in non batch processing. We then break down the message into its payload and
+ * begin validating the payload. Once the payload is validated successfully it will be written to the database to
+ * the relevant tables. If the validation fails it is still written to the database for auditing purposes but with an
+ * INVALID status. For any duplicate requests we will send appropriate callback based on the transfer state and the hash
+ * validation.
+ *
+ * Module.method called to [TODO add description here]
+ *
+ * @param {error} error - error thrown if something fails within Kafka
+ * @param {array} messages - a list of messages to consume for the relevant topic
+ *
+ * @returns {object} - Returns a boolean: true if successful, or throws and error if failed
+ */
+const bulkProcessing = async (error, messages) => {
```

Comment:
This is some pretty hefty business logic here - I didn't notice any new tests for this function. Are they forthcoming? Or are they elsewhere that I've already missed?

Reply:
Actually this logic was designed and implemented in order to shorten/simplify the bulk processing handler code. You might be able to understand it more easily if you review the [sequence diagram](https://github.com/mojaloop/documentation/blob/master/mojaloop-technical-overview/central-bulk-transfers/assets/diagrams/sequence/seq-bulk-1.4.1-bulk-processing-handler.plantuml). Unit and integration tests are to be done after the POC as a part of the prep for merging the poc branch to master.

Reply:
Cool. Thanks for the heads up.